File: 3943-number-of-pairs-after-increment/3943-number-of-pairs-after-increment.java
class Solution {

    public int[] numberOfPairs(int[] nums1, int[] nums2, int[][] q) {

        int n = nums2.length;
        int len = (int)Math.sqrt(n) + 1;

        long[] arr = new long[n];

        for(int i = 0; i < n; i++) {
            arr[i] = nums2[i];
        }

        ArrayList<HashMap<Long,Integer>> mp = new ArrayList<>();

        for(int i = 0; i < len; i++) {
            mp.add(new HashMap<>());
        }

        long[] lazy = new long[len];

        for(int i = 0; i < n; i++) {
            HashMap<Long,Integer> cur = mp.get(i / len);
            cur.put(arr[i], cur.getOrDefault(arr[i], 0) + 1);
        }

        ArrayList<Integer> ans = new ArrayList<>();

        for(int[] qq : q) {

            // update query
            if(qq.length == 4) {

                int l = qq[1];
                int r = qq[2];
                long val = qq[3];

                while(l <= r && l % len != 0) {

                    int b = l / len;

                    HashMap<Long,Integer> cur = mp.get(b);

                    long oldVal = arr[l];

                    cur.put(oldVal, cur.get(oldVal) - 1);

                    if(cur.get(oldVal) == 0) {
                        cur.remove(oldVal);
                    }

                    arr[l] += val;

                    cur.put(arr[l], cur.getOrDefault(arr[l], 0) + 1);

                    l++;
                }

                while(l <= r && (r + 1) % len != 0) {

                    int b = r / len;

                    HashMap<Long,Integer> cur = mp.get(b);

                    long oldVal = arr[r];

                    cur.put(oldVal, cur.get(oldVal) - 1);

                    if(cur.get(oldVal) == 0) {
                        cur.remove(oldVal);
                    }

                    arr[r] += val;

                    cur.put(arr[r], cur.getOrDefault(arr[r], 0) + 1);

                    r--;
                }

                while(l <= r) {
                    lazy[l / len] += val;
                    l += len;
                }

            } else {

                long tot = qq[1];

                int cnt = 0;

                for(int x : nums1) {

                    long needBase = tot - x;

                    for(int b = 0; b < mp.size(); b++) {

                        HashMap<Long,Integer> cur = mp.get(b);

                        cnt += cur.getOrDefault(needBase - lazy[b], 0);
                    }
                }

                ans.add(cnt);
            }
        }

        int[] res = new int[ans.size()];

        for(int i = 0; i < ans.size(); i++) {
            res[i] = ans.get(i);
        }

        return res;
    }
}

File: 3943-number-of-pairs-after-increment/3943-number-of-pairs-after-increment.cpp
class Solution {
public:
using ll=long long;
    vector<int> numberOfPairs(vector<int>& nums1, vector<int>& nums2, vector<vector<int>>& q) {
         int n=nums2.size();
        int len=ceil(sqrt(n));
        vector<ll>arr(nums2.begin(),nums2.end());
        vector<unordered_map<ll,ll>>mp(len);
        vector<ll>lazy(len,0);
         
       for(int i=0;i<arr.size();i++){
           mp[i/len][arr[i]]++;
       }
        
        vector<int>ans;

        for(int i=0;i<q.size();i++){
              if(q[i].size()==4){
                     //update query
                     int l=q[i][1],r=q[i][2];
                     ll val=q[i][3];

                     while(l%len!=0 && l<=r){
                        mp[l/len][arr[l]]--;
                        arr[l]=arr[l]+val;
                        mp[l/len][arr[l]]++;
                        l++;
                     }

                     while((r+1)%len!=0 && l<=r){
                        mp[r/len][arr[r]]--;
                        arr[r]=arr[r]+val;
                        mp[r/len][arr[r]]++;
                        r--;
                     }
                     if(l<r){
                     for(int j=l/len;j<=r/len;j++){
                        lazy[j]+=val;
                     }
                     }
              } else{
                    
                    int cnt=0;
                    int tot=q[i][1];
                    for(int j=0;j<nums1.size();j++){
                        int idx=0;
                         for(auto &e:mp){
                            if(e.find(tot-nums1[j]-lazy[idx])!=e.end()) cnt+=e[(tot-nums1[j]-lazy[idx])];
                            idx++;
                         }
                    }
                    ans.push_back(cnt);
              }
        }
        
        return ans;
    }
};